fix: prevent nil pointer dereference in RemoveImageFromRepoMeta

M0Rf30 opened the pull request:
## Summary

This PR fixes a critical bug where `RemoveImageFromRepoMeta` crashes with a nil pointer dereference during retention policy execution and GC operations.

## Root Cause

The function was accessing blob metadata without checking if it exists first. During GC/retention operations, the metadata database might have stale references to blobs that no longer exist, causing runtime panics at `pkg/meta/common/common.go:279` and `pkg/meta/common/common.go:290-297`.

## Changes

- Added nil check for `descriptorBlobInfo` before accessing `LastUpdated` field (line 279)
- Added nil check for `blobInfo` before dereferencing `Size`, `Vendors`, `Platforms`, and `SubBlobs` (line 296)
- Made the function consistent with `recalculateAggregateFields` which already had these checks

## Impact

This fix resolves:
- Runtime crashes during retention policy execution
- Crashes during GC manifest removal
- Image deletion failures via API
- Eliminates need for `dryRun: true` workaround in retention config

## Testing

- All existing tests pass
- The fix is consistent with similar functions in the codebase
- Code compiles without errors

The fix gracefully handles missing blob metadata by skipping those entries instead of crashing.

## Related Issues

Fixes issue similar to #3273 (same crash pattern in earlier version)

rchincha commented:
> We should understand how we got into this situation in the first place, but the fix looks good.

I agree. While the fix looks good, the worry is we are masking over root cause.

| for tag, descriptor := range repoMeta.Tags {

Maybe there is a path where we are not updating the repoMeta whenever we delete blobs - perhaps GC or retention policy paths.

rchincha commented:
| Area      | File & Function                                                        | Main Risk Location                                                 |
|-----------|-----------------------------------------------------------------------|-------------------------------------------------------------------|
| GC        | [`pkg/storage/gc/gc.go`](https://github.com/project-zot/zot/blob/main/pkg/storage/gc/gc.go), `removeUnreferencedBlobs` | Looping over allBlobs and accessing blob metadata fields          |
| GC        | [`pkg/storage/gc/gc.go`](https://github.com/project-zot/zot/blob/main/pkg/storage/gc/gc.go), `gcManifest`            | Removing manifest if canGC, possible nil if blob metadata missing |
| GC        | [`pkg/storage/gc/gc.go`](https://github.com/project-zot/zot/blob/main/pkg/storage/gc/gc.go), `addIndexBlobsToReferences`, `addImageIndexBlobsToReferences` | Accessing referenced blobs that may be missing                    |
| Retention | [`pkg/retention/retention.go`](https://github.com/project-zot/zot/blob/main/pkg/retention/retention.go), `GetRetainedTagsFromMetaDB` | Accessing candidate statistics that might not exist               |
| Retention | [`pkg/retention/retention.go`](https://github.com/project-zot/zot/blob/main/pkg/retention/retention.go), `groupCandidatesByTagPolicy` and related logic | Grouping/tag logic using possibly missing metadata                |

M0Rf30 commented:
Thanks @copilot-pull-request-reviewer for the thorough review! I've addressed all three feedback items in commit 433c21fa:

### 1. Fixed blob count expectation (line 196)
- Added `delete(repoMeta.Tags, "tag1")` before calling `RemoveImageFromRepoMeta`
- Changed expectation from 2 blobs to 0 blobs (since no tags remain after removal)

### 2. Added tag removal verification (line 240)
- Added assertion: `So(resultMeta.Tags["tag1"], ShouldBeNil)` to verify the tag was removed from metadata

### 3. Fixed blob count and assertions (line 245)
- Added `delete(repoMeta.Tags, "tag1")` before calling the function
- Updated expectation from 4 blobs to 2 blobs (only manifest2 and layer2 from tag2)
- Updated size calculation to match: `int64(2000 + 800)` instead of all 4 blobs

All tests now correctly reflect the actual usage pattern where the caller must delete the tag from `repoMeta.Tags` **before** calling `RemoveImageFromRepoMeta`, as demonstrated in `pkg/meta/boltdb/boltdb.go:1436`.

M0Rf30 commented on pkg/meta/common/common_test.go:
```diff
@@ -126,4 +128,178 @@ func TestUtils(t *testing.T) {
+				},
+			}
+
+			resultMeta, resultBlobs := common.RemoveImageFromRepoMeta(repoMeta, repoBlobs, "tag1")
```
Fixed in commits 433c21fa and 24a09b9f! Both test cases now include the assertion:
- Line 200: `So(resultMeta.Tags["tag1"], ShouldBeNil)`
- Line 253: `So(resultMeta.Tags["tag1"], ShouldBeNil)`

The tag is properly deleted before calling the function and verified afterwards.

M0Rf30 commented on pkg/meta/common/common_test.go:
```diff
@@ -126,4 +128,187 @@ func TestUtils(t *testing.T) {
+			So(resultMeta, ShouldNotBeNil)
+			So(resultBlobs, ShouldNotBeNil)
+			So(resultMeta.LastUpdatedImage, ShouldBeNil)
+		})
```
Revised in commit 6276f1e6! The test now covers the more realistic scenario:

**Before:** Only tested tag1 with missing blob info (unrealistic since we're removing it anyway)

**After:** 
- tag1 has valid blob info and is being removed (deleted from repoMeta.Tags first)
- tag2 remains in repoMeta.Tags but has missing blob info (simulating data inconsistency)
- Verifies the function doesn't panic and completes successfully
- Confirms tag2 remains in metadata but has no blobs (acceptable inconsistent state in GC scenarios)

This properly tests the nil check's ability to handle data inconsistencies for remaining tags during removal operations.

rchincha commented:
@M0Rf30 

| Commits must have verified signatures.

Pls sign and sign-off your commit.

`git amend --signoff -s`
